sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp

```diff
@@ -0,0 +1,341 @@
+
+  template <template <typename T> class PtrT, typename T>
+  enable_if_t<std::is_convertible<PtrT<T>, weak_ptr_class<T>>::value>
+  set_final_data(PtrT<T> FinalData) {
+    weak_ptr_class<T> TempFinalData(FinalData);
+    set_final_data(TempFinalData);
+  }
+
+  template <typename T> void set_final_data(weak_ptr_class<T> FinalData) {
+    MUploadDataFunctor = [this, FinalData]() {
+      if (shared_ptr_class<T> LockedFinalData = FinalData.lock()) {
+        EventImplPtr Event = updateHostMemory(LockedFinalData.get());
+        if (Event)
+          Event->wait(Event);
+      }
+    };
+  }
+
+  void set_final_data_from_storage() {
+    MUploadDataFunctor = [this]() {
+      if (!MSharedPtrStorage.unique()) {
+        void *FinalData = const_cast<void *>(MSharedPtrStorage.get());
+        EventImplPtr Event = updateHostMemory(FinalData);
+        if (Event)
+          Event->wait(Event);
+      }
+    };
+  }
+
+  template <typename Destination>
+  EnableIfOutputPointerT<Destination> set_final_data(Destination FinalData) {
+    MUploadDataFunctor = [this, FinalData]() {
+      EventImplPtr Event = updateHostMemory(FinalData);
+      if (Event)
+        Event->wait(Event);
+    };
+  }
+
+  template <typename Destination>
+  EnableIfOutputIteratorT<Destination> set_final_data(Destination FinalData) {
+    MUploadDataFunctor = [this, FinalData]() {
+      using DestinationValueT = iterator_value_type_t<Destination>;
+      // TODO if Destination is ContiguousIterator then don't create
+      // ContiguousStorage. updateHostMemory works only with pointer to
+      // continuous data.
+      const size_t Size = MSizeInBytes / sizeof(DestinationValueT);
+      vector_class<DestinationValueT> ContiguousStorage(Size);
+      EventImplPtr Event = updateHostMemory(ContiguousStorage.data());
+      if (Event) {
+        Event->wait(Event);
+        std::copy(ContiguousStorage.cbegin(), ContiguousStorage.cend(),
+                  FinalData);
+      }
+    };
+  }
+
+  EventImplPtr updateHostMemory(void *const Ptr) {
+    const id<3> Offset{0, 0, 0};
+    const range<3> AccessRange{MSizeInBytes, 1, 1};
+    const range<3> MemoryRange{MSizeInBytes, 1, 1};
+    const access::mode AccessMode = access::mode::read;
+    SYCLMemObjI *SYCLMemObject = this;
+    const int Dims = 1;
+    const int ElemSize = 1;
+
+    Requirement Req(Offset, AccessRange, MemoryRange, AccessMode, SYCLMemObject,
+                    Dims, ElemSize);
+    Req.MData = Ptr;
+
+    EventImplPtr Event = Scheduler::getInstance().addCopyBack(&Req);
+    return Event;
+  }
+
+  // Update host with the latest data + notify scheduler that the memory object
+  // is going to die. After this method is finished no further operations with
+  // the memory object is allowed. This method is executed from child's
+  // destructor. This cannot be done in SYCLMemObjT's destructor as child's
+  // members must be alive.
+  void writeBack() {
```

Comment:
```suggestion
  void updateHostMemory() {
```

Thanks for the comment.